client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerLeakTest.java

```diff
@@ -38,7 +41,10 @@ public void call() {
             eventListener.expectOnlyCreatedEvent(1, cache(cacheName));
          }
       });
-      detectThreadLeaks(".*Client-Listener-ClientListenerLeakTest-.*");
+      WaitFor
+              .condition(() -> listThreadsMatchingRegexp(".*Client-Listener-ClientListenerLeakTest-.*").size() == 0)
+              .atMost(10, TimeUnit.MILLISECONDS)
+              .await();
```

Comment:
Let me answer your comments before proposing solution:

- Unfortunately not all caught threads are in `TERMINATED` state. Some of them are in `RUNNABLE` state and they were waiting for `Unsafe.unpark` method (at least this happens on my JVM `Java version: 1.8.0_40, vendor: Oracle Corporation`). So the simplest possible trick - adding `Thread.isAlive` to `if (thread.getName().matches(regexp))` didn't work. 
- I think `shutdown()` is fine there... I noticed that the test doesn't wait for terminating clients. It's more about the method that we use to detect leaked threads. 

Looking at your proposed changes, I have implemented another fix. In order to decouple reverting the original thread name and the work to be done - I used `ThreadFactory` interface and passed it to the executor.